azure/services/agentpools/spec.go

```diff
@@ -352,7 +352,7 @@ func (s *AgentPoolSpec) Parameters(ctx context.Context, existing interface{}) (p
 			VnetSubnetID:         vnetSubnetID,
 			EnableNodePublicIP:   s.EnableNodePublicIP,
 			NodePublicIPPrefixID: s.NodePublicIPPrefixID,
-			Tags:                 *azure.StringMapPtr(s.AdditionalTags),
+			Tags:                 converters.TagsToMap(s.AdditionalTags),
```

Comment:
It seems that this change is what is causing the aks e2e test to fail:

```
The function passed to Eventually failed at /home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/aks_tags.go:182 with:
Expected
    <v1beta1.Tags | len:2>: {"new": "value", "test": "updated"}
to be nil failed [FAILED] Timed out after 1800.000s.
The function passed to Eventually failed at /home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/aks_tags.go:182 with:
Expected
    <v1beta1.Tags | len:2>: {"new": "value", "test": "updated"}
to be nil
In [It] at: /home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/aks_tags.go:229 @ 03/17/23 15:25:09.421
```

Reply:
@nojnhuh any ideas on this one? I think that's an e2e test you added recently, want to make sure we're not regressing things here

Reply:
I think in this case we wanted the behavior where a nil `s.AdditionalTags` should result in the `Tags` being set to a non-nil, empty map which tells AKS to delete all the tags which is what the test is trying to do. Sending nil to the API means "don't change the tags" which seems to be what the test is running into.

Reply:
Got it, I wonder if we should just change this to 

```
		Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
			Lifecycle:   infrav1.ResourceLifecycleOwned,
			ClusterName: s.ClusterName,
			Name:        pointer.String(s.Name),
			Role:        pointer.String(infrav1.CommonRole),
			Additional:  s. AdditionalTags,
		})),
```

then, just like you did in https://github.com/CecileRobertMichon/cluster-api-provider-azure/commit/ec154f70c636aca350990a81e980d9b4d7ebdb11#diff-d7496d9845d55dda5142aab2970a5a9aa1012fbb52d2c462cc5b220682ca520a so we're also tagging agentpools with the standard CAPZ owned tags?

Reply:
FWIW I didn't see tags on agent pools affected by the same constant reconciliation issue as the managed cluster resource from #3009, so it seems like tags on agent pools are classified differently somehow. That seems like a reasonable change to add those labels though since I don't see how that would hurt anything. I can put that change together quick and assuming that goes smoothly, this PR can maybe just ignore the whole issue.

Reply:
> assuming that goes smoothly

It did not :/ 

AKS doesn't seem to let you use the same "update the tags for this resource ID" API for agent pools. Everything looks like it works, but the tags are applied to the managed cluster, not the agent pool.

I thought I may have made a mistake wiring things up in CAPZ, but the CLI seems to work the same way:
```
% az tag update --resource-id /subscriptions/$AZURE_SUBSCRIPTION_ID/resourceGroups/aks-23760/providers/Microsoft.ContainerService/managedClusters/aks-23760 --operation merge --tags jontocluster=fromcli

% az tag update --resource-id /subscriptions/$AZURE_SUBSCRIPTION_ID/resourceGroups/aks-23760/providers/Microsoft.ContainerService/managedClusters/aks-23760/agentPools/pool0 --operation merge --tags jontopool=fromcli

% az tag list --resource-id /subscriptions/$AZURE_SUBSCRIPTION_ID/resourceGroups/aks-23760/providers/Microsoft.ContainerService/managedClusters/aks-23760/agentPools/pool0 -o yaml
id: /subscriptions/$AZURE_SUBSCRIPTION_ID/resourceGroups/aks-23760/providers/Microsoft.ContainerService/managedClusters/aks-23760/providers/Microsoft.Resources/tags/default
name: default
properties:
  tags:
    Name: aks-23760
    jontocluster: fromcli
    jontopool: fromcli
    sigs.k8s.io_cluster-api-provider-azure_cluster_aks-23760: owned
    sigs.k8s.io_cluster-api-provider-azure_role: common
resourceGroup: aks-23760
type: Microsoft.Resources/tags
```
^ The `id` from the output refers to the managed cluster.

I think getting this to work just like tags on managed cluster resources will be more complex than it might be worth, so I'm in favor of adding the minimal logic here to pass the e2e test at least for the short-term.

Reply:
> so I'm in favor of adding the minimal logic here to pass the e2e test at least for the short-term
 
which minimal logic?

Reply:
Makes sense to me. @PiotrLewandowski323 let us know if you need additional clarification on this. 

This would look something like this:

```
tags := converters.TagsToMap(s.AdditionalTags)
if tags == nil {
  // Make sure we send a non-nil, empty map if AdditionalTags are nil as this tells AKS to delete any existing tags.
  tags = make(map[string]*string, 0)
}

...

Tags:                 tags,

```

Reply:
@nojnhuh one more question for my own understanding: if we set the agentpool tags to an empty map when AdditionalTags is nil, wouldn't that result in endless reconciliation if some other process is adding tags to the agentpools in the sub?

Specifically, this part: 

> I didn't see tags on agent pools affected by the same constant reconciliation issue as the managed cluster resource from https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3009

Reply:
I think it would, but at least the "add a tag to all resources" Azure Policy I found to test that issue didn't seem to apply to agent pools, just the managed cluster. That makes me wonder how easy it would be to get into that scenario in the first place. It seems like it would have to be very deliberate.